### ISSUE TITLE
Q4 goal: Integrate BinderHub into JupyterHub (funded by GESIS)

### ISSUE THREAD

**Issue by yuvipanda** [2i2c-org/infrastructure#3286]
# Technical Progress

These are broad outcomes of technical progress that need to happen for smoother UX. We currently have code that is decent demo quality, but needs to get closer to production quality

- [x] Complete productionizing `@jupyterhub/binderhub-client` JS library and release on npmjs, so we can use the same code as binderhub for talking to the binderhub API
- [x] Get jupyterhub-fancy-profiles to a first release, including deployment on at least one other hub that does not use binderhub building (to make sure it is a broadly useful project)
- [ ] Make sure auth works ok, so arbitrary users who are not authenticated with the JupyterHub can not use the binderhub API to autobuildi

## Out of scope items

Since work will continue in Q1, the following items are explicitly marked as out of scope for Q4.

- URL compatibility with mybinder.org, so users can more easily shift
- Options for figuring out how to expose the content of the repository to the end user on their home directory in a persistent way. 
- Long term planning for where `binderhub-service` and `jupyter-fancy-profiles` repo live. There is ongoing conversation in the jupyterhub community about a `jupyterhub-contrib` org, so let's wait on that.

# Deploy this on non-2i2c infrastructure

We need to make sure that all this work is not 2i2c specific. While the tech work has proceeded in a non-2i2c specific way, we currently don't have any deployment of this outside 2i2c infrastructure. This means there isn't clear step by step documentation on tying together the three projects (z2jh, binderhub-service, jupyterhub-fancy-profiles) that make this workflow possible.

We work potentially with Sima Attar to get a deployment running elsewhere - either on GCP (we have funding to setup a spare project) or on other infrastructure they have access to. The outcome of this is documentation that can be followed by anyone to set this up.

# Write a blog post about this complex yet 'upstream first' process

There's been a lot of work done in making this goal possible, spread out across multiple projects - `kubespawner`, `binderhub`, `zero-to-jupyterhub-on-k8s`, `repo2docker`, `jupyterhub-fancy-profiles`, etc. For organizations approaching contributing major features to an open source ecosystem, this can be quite daunting! So we write a blog post describing how we did this, including the exporation paths that didn't go anywhere.

**Comment by rgaiacs** [comment]
Hi @yuvipanda, I might be able to help with the deploy on non-2i2c infrastructure. Can you point me to where I can start?

**Comment by yuvipanda** [comment]
I had a nice conversation with Sima a few days ago, and we're working on getting this all set up on GCP. I've given her access to the earlier GCP project we used for testing (`binderhub-service-development` I think), and are going to take it slow :)

**Comment by yuvipanda** [comment]
Would love for you to try it out too, @rgaiacs

**Comment by arnim** [comment]
We probably should also add documentation, for users and admins. Having documentation is also relevant for the NFDI and CESSDA to allow them to deploy this.

**Comment by yuvipanda** [comment]
Some next steps from chatting with @arnim:

1. Write a 'status report' blog post for the *2i2c blog*, with a focus on all the upstream work that has happened: https://hackmd.io/hiPdQpL7ScC6_IHopInf0A
2. Figure out what would be *useful* to post to the Jupyter blog - particularly, the fact that we had to thread the needle via many parts of the Jupyter ecosystem to get this to happen.
3. Cleanup some issues for a demo @arnim is giving on Nov 23.

**Comment by arnim** [comment]
Thank you @yuvipanda . 
These next step are perfect 👍

**Comment by yuvipanda** [comment]
As an update, Sima has spent some time getting acquainted with GCP and set up a z2jh there to test (https://2i2c.slack.com/archives/C03RLNFM43F/p1699262794125269).

Next step is to try deploy binderhub-service there alongside.

**Comment by arnim** [comment]
Just had a meeting with @simaattar2003 on the deployment.

**Comment by yuvipanda** [comment]
We've done a lot this quarter!

- We published https://2i2c.org/blog/2024/jupyterhub-binderhub-gesis/, which lists all the work that has happened through this grant. Thanks to everyone for their review!
- We have a working demo at imagebuilding-demo.2i2c.cloud, and folks have been testing it out.
- There's now a [npm package](https://www.npmjs.com/package/@jupyterhub/binderhub-client) for binderhub-client that's upstream
- I've been working with @simaattar2003 via video calls, creating documentation for recreating this setup elsewhere. This is ongoing work.

I'm going to close this out, and create a new issue for Q1 2024. Thanks everyone.
